### ISSUE TITLE
WeakHashSet::retain doesn't iterate over all valid items

### ISSUE THREAD

**Issue by m-demare** [tov/weak-table-rs#22]
Hi, I'm having this issue where *sometimes* WeakHashSet::retain skips some valid items.

For example, the following code fails about 1 out of 5 times I run it (I think the randomness is probably related to the randomness in a HashSet's iteration order):

```rust
use std::{
    fmt::Write,
    rc::{Rc, Weak},
};

use weak_table::WeakHashSet;

fn main() {
    let mut set = WeakHashSet::default();
    let mut v = Vec::new();

    let n = 5;

    for i in 0..n {
        let rc = Rc::new(i);
        v.push(rc.clone());
        set.insert(rc);
    }

    let rc = Rc::new(n);
    set.insert(rc.clone());

    print_set(&set, "before drop");

    drop(v);

    print_set(&set, "after drop");

    let mut retain_calls = Vec::new();
    set.retain(|val| {
        retain_calls.push(val);
        false
    });

    assert!(
        retain_calls.iter().any(|v| v == &n.into()),
        "retain was not called on {n}"
    );
}

fn print_set(set: &WeakHashSet<Weak<u32>>, name: &str) {
    let elems = set.iter().fold(String::new(), |mut f, val| {
        let _ = write!(f, "{val}, ");
        f
    });
    println!("elems ({name}): {elems}");
}
```

If you need any further info, please tell me. Thanks for making this incredibly useful crate!

**Comment by nmathewson** [comment]
Not sure if this crate is still maintained, but I've looked at the bug a bit to diagnose the issue and to see whether any other APIs are affected.  Here's my theory:

The `retain` method works by calling `remove_index`, which potentially shifts elements within the hash table in order to preserve invariants.  (The hash table uses open addressing, so whenever an element is removed,  we need to check whether another element should go into the now-vacated bucket.)  But any elements that shift as part of this `remove_index` call are not automatically checked in the `retain` loop.

**Comment by m-demare** [comment]
> The `retain` method works by calling `remove_index`, which potentially shifts elements within the hash table in order to preserve invariants. (The hash table uses open addressing, so whenever an element is removed, we need to check whether another element should go into the now-vacated bucket.) But any elements that shift as part of this `remove_index` call are not automatically checked in the `retain` loop.

Yeah, that sounds right. Sadly it feels like the crate is unmantained, so I switched to using a `Vec<Weak<T>>` which is good enough for my usecase. I use:

```rust
vec.retain(|el| match el.upgrade() {
    Some(el) => {
        should_retain(el)
    }
    None => false,
});
```

for retaining

and

```rust
if vec.len() == vec.capacity() {
    vec.retain(|el| el.upgrade().is_some());
    vec.reserve(vec.len()); // Reserve at least n more slots
}
vec.push(val);
```

to push an element and remove invalid ones (keeping it O(1) amortized)

**Comment by nmathewson** [comment]
> Yeah, that sounds right. Sadly it feels like the crate is unmantained, so I switched to using a `Vec<Weak<T>>` which is good enough for my usecase. 

For what it's worth, I've started a `weak-table2` that uses a different backend, and doesn't have this bug.  I'm currently pausing on it while I wait to see if anybody wants to use it, and if @tov wants it to become the next version of `weak-table` .

**Comment by m-demare** [comment]
Cool! I'll try it

**Comment by nmathewson** [comment]
Update: @tov has graciously given me permission to work on this crate, so I'll be putting out an 0.4.0 soon with this bug fixed, and several other improvements.

**Comment by nmathewson** [comment]
I've put out 0.4.0-pre1 and 0.4.0-pre2 with the fix for this issue; they should be good to use.  0.4.0 will be out soon.

**Comment by m-demare** [comment]
Nice, I'm glad to see the crate supported again, thank you!
